rollup/da_syncer/blob_client/beacon_node_client.go

```diff
@@ -0,0 +1,183 @@
+	genesisPath, err := url.JoinPath(apiEndpoint, beaconNodeGenesisEndpoint)
+	if err != nil {
+		return nil, fmt.Errorf("failed to join path, err: %w", err)
+	}
+	resp, err := http.Get(genesisPath)
+	if err != nil {
+		return nil, fmt.Errorf("cannot do request, err: %w", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		body, _ := io.ReadAll(resp.Body)
+		bodyStr := string(body)
+		return nil, fmt.Errorf("beacon node request failed, status: %s, body: %s", resp.Status, bodyStr)
+	}
+
+	var genesisResp GenesisResp
+	err = json.NewDecoder(resp.Body).Decode(&genesisResp)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode result into struct, err: %w", err)
+	}
+	genesisTime, err := strconv.ParseUint(genesisResp.Data.GenesisTime, 10, 64)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode genesis time %s, err: %w", genesisResp.Data.GenesisTime, err)
+	}
+
+	// get seconds per slot from spec
+	specPath, err := url.JoinPath(apiEndpoint, beaconNodeSpecEndpoint)
+	if err != nil {
+		return nil, fmt.Errorf("failed to join path, err: %w", err)
+	}
+	resp, err = http.Get(specPath)
+	if err != nil {
+		return nil, fmt.Errorf("cannot do request, err: %w", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		body, _ := io.ReadAll(resp.Body)
+		bodyStr := string(body)
+		return nil, fmt.Errorf("beacon node request failed, status: %s, body: %s", resp.Status, bodyStr)
+	}
+
+	var specResp SpecResp
+	err = json.NewDecoder(resp.Body).Decode(&specResp)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode result into struct, err: %w", err)
+	}
+	secondsPerSlot, err := strconv.ParseUint(specResp.Data.SecondsPerSlot, 10, 64)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode seconds per slot %s, err: %w", specResp.Data.SecondsPerSlot, err)
+	}
+	if secondsPerSlot == 0 {
+		return nil, fmt.Errorf("failed to make new BeaconNodeClient, secondsPerSlot is 0")
+	}
+
+	return &BeaconNodeClient{
+		apiEndpoint:    apiEndpoint,
+		l1Client:       l1Client,
+		genesisTime:    genesisTime,
+		secondsPerSlot: secondsPerSlot,
+	}, nil
+}
+
+func (c *BeaconNodeClient) GetBlobByVersionedHashAndBlockNumber(ctx context.Context, versionedHash common.Hash, blockNumber uint64) (*kzg4844.Blob, error) {
+	// get block timestamp to calculate slot
+	header, err := c.l1Client.GetHeaderByNumber(blockNumber)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get header by number, err: %w", err)
+	}
+	slot := (header.Time - c.genesisTime) / c.secondsPerSlot
+
+	// get blob sidecar for slot
+	blobSidecarPath, err := url.JoinPath(c.apiEndpoint, beaconNodeBlobEndpoint, fmt.Sprintf("%d", slot))
+	if err != nil {
+		return nil, fmt.Errorf("failed to join path, err: %w", err)
+	}
+	resp, err := http.Get(blobSidecarPath)
+	if err != nil {
+		return nil, fmt.Errorf("cannot do request, err: %w", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		body, _ := io.ReadAll(resp.Body)
```

Comment:
what about adding error handling? also in the other 2 client implementations.

Reply:
Done in 6d1e5cec498e29c852abf25a05d5c858f722a562